client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -526,6 +526,7 @@ void reconnect() {
                          try {
                              connection.send(cmd);
                          } catch (ConnectionFailedException e1) {
+                             state.failConnection(e1);
```

Comment:
it is an intuitive change, but I want understand your thought process to come to the conclusion that this line here is needed. I suppose that we are now failing the connection elsewhere while we should, but I want to hear your detailed explanation, please.

Reply:
The following operations are performed by `state.setupConnection.releaser` :
1. Get the endpoint (Segment Store) details of the given segment.
2. Use `connectionFactory.establishConnection()` to establish a connection to the Segment Store.
Once the `io.pravega.client.netty.impl.ClientConnection` is returned by the `establishConnection()` method a `SetupAppend` is send over this connection.

The issue we are observing is that once a client connection is successfully returned by the connectionFactory an `ConnectionFailedException` is received while sending the `SetupAppend`. Due to this exception the reconnect logic is retried causing the Precondition checkstate to fail.
This causes the reconnect logic to be invoked indefinitely there by blocking all operations for this SegmentOutputStream.

By invoking state.failConnection() we are ensuring that the connection is torn down in this error scenario.

Note: `io.netty.channel.ChannelInboundHandlerAdapter#channelRegistered` needs to be invoked for the channel to be not null. In this corner case we are seeing
that `io.netty.bootstrap.Bootstrap#connect(java.lang.String, int)` completed successfully but the `channelRegistered()` method was not invoked causing the send operation to fail.

Reply:
What's the intention of the sneaky throw following this line? It feels like the intention is to break the retry loop, but we are actually retrying unconditionally, so it won't break. Is the right way to fix this not to retry unconditionally? In fact, what happens if the reconnect breaks loop with an exception?

Reply:
`io.pravega.common.util.ReusableFutureLatch` : An exception thrown within the releaser (Runnable)  will cause none of the waiting futures to be notified .

Reply:
This change is not correct. The throw on the line below will already trigger a retry. The problem this is trying to address is that the retry keeps failing becauses close is never being called on the connection so the loop always fails in the same place. However this is an artifact of channelUnregisted not being invoked. Similar to the channelRegistered issue mentioned above, this is likely a result in a change in the way netty handles things. I think it is our fault though. Reading their documentation it looks like we should be overriding the callbacks for channelActive and channelInactive rather than registered and unregistered. (We previously made this change in a PR and ended up backing it out.) I think we should make that change as see if it fixes the problem of the connect and retries when send fails.

Reply:
Updated the code to use callbacks `channelActive, channelInactive` rather than `channelRegistered, channelUnregistered`.